File: 04-Conjuntos/cpp/conjunto.cpp
#include "conjunto.h"

template <typename T>
int Conjunto<T>::length() {
    return data.size();
}

template <typename T>
bool Conjunto<T>::contains(T value) {
    for(int i = 0; i < length(); i++) {
        if(data[i] == value) {
            return true;
        }
    }
    return false;
}

template <typename T>
void Conjunto<T>::add(T value) {
    if(!contains(value)) {
        data.push_back(value);
    }
}

template <typename T>
void Conjunto<T>::remove(T value) {
    for(int i = 0; i < length(); i++) {
        if(data[i] == value) {
            data.erase(data.begin() + i);
            return;
        }
    }
}

template <typename T>
bool Conjunto<T>::equals(Conjunto<T> *other) {
    if(length() != other->length()) {
        return false;
    }
    for(int i = 0; i < length(); i++) {
        if(!other->contains(data[i])) {
            return false;
        }
    }
    return true;
}

template <typename T>
bool Conjunto<T>::isSubset(Conjunto<T> *other) {
    if(length() < other->length()) {
        return false;
    }
    for(int i = 0; i < other->length(); i++) {
        if(!contains(other->data[i])) {
            return false;
        }
    }
    return true;
}

template <typename T>
void Conjunto<T>::union_(Conjunto<T> *other) {
    for(int i = 0; i < other->length(); i++) {
        add(other->data[i]);
    }
}

template <typename T>
Conjunto<T> Conjunto<T>::intersection(Conjunto<T> *other) {
    Conjunto<T> result;
    for(int i = 0; i < length(); i++) {
        if(other->contains(data[i])) {
            result.add(data[i]);
        }
    }
    return result;
}

template <typename T>
Conjunto<T> Conjunto<T>::difference(Conjunto<T> *other) {
    Conjunto<T> result;
    for(int i = 0; i < length(); i++) {
        if(!other->contains(data[i])) {
            result.add(data[i]);
        }
    }
    for(int i = 0; i < other->length(); i++) {
        if(!contains(other->data[i])) {
            result.add(other->data[i]);
        }
    }
    return result;
}

File: 04-Conjuntos/java/src/Conjunto.java
import java.util.ArrayList;

public class Conjunto<T> implements Iterable<T>{
    ArrayList<T> data = new ArrayList<>();

    public int length() {
        return data.size();
    }

    public boolean contains(T value) {
        return data.indexOf(value) != -1;
    }

    public void add(T value) {
        if(!contains(value)) {
            data.add(value);
        }
    }

    public void remove(T value) {
        data.remove(value);
    }

    public boolean equals(Conjunto<T> other) {
        if(length() != other.length()) {
            return false;
        }
        for(int i = 0; i < length(); i++) {
            if(!other.contains(data.get(i))) {
                return false;
            }
        }
        return true;
    }

    public boolean isSubset(Conjunto<T> other) {
        if(length() < other.length()){
            return false;
        }
        for(int i = 0; i < other.length(); i++) {
            if(!contains(other.data.get(i))) {
                return false;
            }
        }
        return true;
    }

    public void union(Conjunto<T> other) {
        for(int i = 0; i < other.length(); i++) {
            add(other.data.get(i));
        }
    }

    public Conjunto<T> intersection(Conjunto<T> other) {
        Conjunto<T> result = new Conjunto();
        for(int i = 0; i < length(); i++) {
            if(other.contains(data.get(i))) {
                result.add(data.get(i));
            }
        }
        return result;
    }

    public Conjunto<T> difference(Conjunto<T> other) {
        Conjunto<T> result = new Conjunto();
        for(int i = 0; i < length(); i++) {
            if(!other.contains(data.get(i))) {
                result.add(data.get(i));
            }
        }
        for(int i = 0; i < other.length(); i++) {
            if(!contains(other.data.get(i))) {
                result.add(other.data.get(i));
            }
        }
        return result;
    }

    @Override
    public String toString() {
        return data.toString();
    }

    @Override
    public java.util.Iterator<T> iterator() {
        return new java.util.Iterator<T>() {
            int index = 0;

            @Override
            public boolean hasNext() {
                return index < data.size();
            }

            @Override
            public T next() {
                return data.get(index++);
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }
}
